crates/bevy_tasks/src/tokio_task_pool.rs

```diff
@@ -0,0 +1,548 @@
+    /// fn scope_escapes_closure() {
+    ///     let pool = TaskPool::new();
+    ///     let foo = Box::new(42);
+    ///     pool.scope(|scope| {
+    ///         std::thread::spawn(move || {
+    ///             // UB. This could spawn on the scope after `.scope` returns and the internal Scope is dropped.
+    ///             scope.spawn(async move {
+    ///                 assert_eq!(*foo, 42);
+    ///             });
+    ///         });
+    ///     });
+    /// }
+    /// ```
+    ///
+    /// ```compile_fail
+    /// use bevy_tasks::TaskPool;
+    /// fn cannot_borrow_from_closure() {
+    ///     let pool = TaskPool::new();
+    ///     pool.scope(|scope| {
+    ///         let x = 1;
+    ///         let y = &x;
+    ///         scope.spawn(async move {
+    ///             assert_eq!(*y, 1);
+    ///         });
+    ///     });
+    /// }
+    ///
+    pub fn scope<'env, F, T>(&self, f: F) -> Vec<T>
+    where
+        F: for<'scope> FnOnce(&'scope Scope<'scope, 'env, T>),
+        T: Send + 'static,
+    {
+        // SAFETY: This safety comment applies to all references transmuted to 'env.
+        // Any futures spawned with these references need to return before this function completes.
+        // This is guaranteed because we drive all the futures spawned onto the Scope
+        // to completion in this function. However, rust has no way of knowing this so we
+        // transmute the lifetimes to 'env here to appease the compiler as it is unable to validate safety.
+        let runtime: &Runtime = &self.runtime;
+        let runtime: &'env Runtime = unsafe { mem::transmute(runtime) };
+        let task_scope_runtime: Runtime = Builder::new_current_thread().build().unwrap();
+        let task_scope_runtime: &'env Runtime =
+            unsafe { mem::transmute(&task_scope_runtime) };
+        let spawned: ConcurrentQueue<JoinHandle<T>> = ConcurrentQueue::unbounded();
+        let spawned_ref: &'env ConcurrentQueue<JoinHandle<T>> =
+            unsafe { mem::transmute(&spawned) };
+
+        let scope = Scope {
+            runtime,
+            task_scope_runtime,
+            spawned: spawned_ref,
+            scope: PhantomData,
+            env: PhantomData,
+        };
+
+        let scope_ref: &'env Scope<'_, 'env, T> = unsafe { mem::transmute(&scope) };
+
+        f(scope_ref);
+
+        if spawned.is_empty() {
+            Vec::new()
+        } else {
+            let get_results = async {
+                let mut results = Vec::with_capacity(spawned_ref.len());
+                while let Ok(task) = spawned_ref.pop() {
+                    results.push(task.await.unwrap());
+                }
+
+                results
+            };
+
+            // Pin the futures on the stack.
+            pin!(get_results);
+
+            let _guard = task_scope_runtime.enter();
+            loop {
+                if let Some(result) = self.runtime.block_on(future::poll_once(&mut get_results)) {
```

Comment:
would a poll_x_times function work? where it polls the future x times before returning ready.

Reply:
tried poking at this a little bit. I think it's getting stuck when it tries to await a local task inside get_results. My current theory is that the task.await never returns, because the executor can't run the task. Currently trying to think of how to confirm this theory.

Reply:
got bevy_tasks tests passing, but running into an issue with par_for_each and tokio not allowing executors to be nested.

https://github.com/james7132/bevy/compare/tokio-task-pool...hymm:bevy:tokio

Reply:
My branch now can run bevy, but it's put together with duck tape. There's some failing tests still on bevy_tasks.

Checking the perf on many_foxes has it running slightly slower. But not sure how much of that is due to my implementation of scope.

![image](https://user-images.githubusercontent.com/2180432/204123055-6e01668a-83c2-490f-8088-10cc195e9954.png)